File: presentations/2021-01-29-graph-dbs-for-code-analysis/graphdb-talk-code.java

  // in class java.sql.Statement
  public boolean execute(String sql); 

  String sql = "drop from users;";
  statement.execute(sql);



       String sql = "delete from users";
       statement.execute(sql);

       String sql = servlet.inputParameter("sql");
       statement.execute(sql);


        public String concatenate(String a, String b) {
          return a + b;
        }
        String select = "select * from user where name = ";
        String user   = servlet.inputParameter("user");
        String sql    = concatenate(select, user);
        statement.execute(sql);

File: presentations/2021-01-29-graph-dbs-for-code-analysis/graphdb-talk-code.scala
  graph.nodes("METHOD")
       .has("name", "Statement.execute")
       .in("TARGET") // -> CALL
       .out("ARGUMENT")

  // same with domain specific traversals (embedded dsl)
  graph.method
   .name("Statement.execute")
   .calledBy
   .argument


  graph.nodes("METHOD")
       .has("name", "Statement.execute")
       .in("TARGET")     // -> CALL
       .out("ARGUMENT")
       .hasLabel("CALL") // only CALLs
       .out("TARGET")    // -> METHOD

  // same with domain specific traversals (embedded dsl)
  graph.method
   .name("Statement.execute")
   .calledBy
   .argument
   .isCall  
   .target  

//with repeat:
  graph.nodes("METHOD")
    .has("name", "Statement.execute")
    .in("TARGET") // -> CALL
    .repeat(
      _.out("ARGUMENT").hasLabel("CALL"))
    (_.until(_.out("TARGET").has("name", "Servlet.inputParameter")))


  // same with domain specific traversals (embedded dsl)
  graph.method.name("java.sql.Statement.execute")
    .calledBy
    .repeat(_.argument.isCall)(
      _.until(_.target.name("Servlet.inputParameter"))
